File: src/scheduler.js
const cron = require('node-cron');
const { execSync } = require('child_process');
const path = require('path');
const winston = require('winston');
const fs = require('fs').promises;

// Load configuration
const configPath = path.join(__dirname, '..', 'config.json');
const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));

// Setup logging
const logger = winston.createLogger({
  level: config.logging.level,
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.json()
  ),
  transports: [
    new winston.transports.File({ 
      filename: path.join(__dirname, '..', config.logging.file) 
    }),
    new winston.transports.Console({
      format: winston.format.simple()
    })
  ]
});

class GoWildScheduler {
  constructor() {
    this.job = null;
    this.isRunning = false;
  }

  async runSearch() {
    if (this.isRunning) {
      logger.warn('Search already in progress, skipping');
      return;
    }

    this.isRunning = true;
    logger.info('Starting scheduled GoWild search...');

    try {
      const searcherPath = path.join(__dirname, 'searcher.js');
      execSync(`node "${searcherPath}"`, {
        stdio: 'inherit',
        timeout: 3600000 // 1 hour timeout
      });
      logger.info('Scheduled search completed successfully');
    } catch (error) {
      logger.error(`Scheduled search failed: ${error.message}`);
      
      // Send error notification
      try {
        const errorMsg = `❌ *GoWild Search Failed*\n\nError: ${error.message}\nTime: ${new Date().toISOString()}`;
        execSync(`openclaw message send --target "${config.slack.channel}" --message "${errorMsg.replace(/"/g, '\\"')}"`);
      } catch (notifyError) {
        logger.error(`Failed to send error notification: ${notifyError.message}`);
      }
    } finally {
      this.isRunning = false;
    }
  }

  start() {
    // Schedule to run daily at 6:00 AM Mountain Time
    // Cron format: second minute hour day month weekday
    // Using 0 6 * * * for 6 AM daily
    this.job = cron.schedule('0 6 * * *', () => {
      this.runSearch();
    }, {
      timezone: 'America/Denver',
      scheduled: true
    });

    logger.info('GoWild scheduler started - runs daily at 6:00 AM MDT');
  }

  stop() {
    if (this.job) {
      this.job.stop();
      logger.info('GoWild scheduler stopped');
    }
  }

  async runNow() {
    logger.info('Running immediate search...');
    await this.runSearch();
  }
}

// CLI interface
async function main() {
  const scheduler = new GoWildScheduler();
  const args = process.argv.slice(2);

  if (args.includes('--start') || args.includes('-s')) {
    scheduler.start();
    console.log('✅ Scheduler started. Press Ctrl+C to stop.');
    
    // Keep process running
    process.on('SIGINT', () => {
      scheduler.stop();
      console.log('\n👋 Scheduler stopped');
      process.exit(0);
    });
    
  } else if (args.includes('--run-now') || args.includes('-r')) {
    await scheduler.runNow();
    scheduler.stop();
    
  } else if (args.includes('--help') || args.includes('-h')) {
    console.log(`
GoWild Fare Scheduler

Usage:
  node scheduler.js [options]

Options:
  --start, -s      Start the daily scheduler (runs at 6 AM MDT)
  --run-now, -r    Run search immediately
  --help, -h       Show this help message

To run as a background service, consider using:
  - nohup: nohup node scheduler.js --start &
  - systemd: Create a service file
  - pm2: pm2 start scheduler.js --name gowild -- --start
    `);
  } else {
    // Default: run once
    await scheduler.runNow();
  }
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { GoWildScheduler };

File: gowild-searcher/scheduler.py
#!/usr/bin/env python3
"""
GoWild Fare Scheduler
Runs the searcher daily at scheduled times
"""

import schedule
import time
import logging
import subprocess
import sys
from pathlib import Path
from datetime import datetime

# Setup paths
BASE_DIR = Path(__file__).parent
LOG_DIR = BASE_DIR / 'logs'
LOG_DIR.mkdir(exist_ok=True)

# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(LOG_DIR / 'scheduler.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)


def run_search():
    """Run the GoWild searcher"""
    logger.info("Starting scheduled GoWild search...")
    
    try:
        searcher_script = BASE_DIR / 'gowild_searcher.py'
        result = subprocess.run(
            [sys.executable, str(searcher_script)],
            capture_output=True,
            text=True,
            timeout=3600  # 1 hour timeout
        )
        
        if result.returncode == 0:
            logger.info("Scheduled search completed successfully")
            if result.stdout:
                logger.info(result.stdout)
        else:
            logger.error(f"Search failed: {result.stderr}")
            send_error_notification(result.stderr)
            
    except subprocess.TimeoutExpired:
        logger.error("Search timed out after 1 hour")
        send_error_notification("Search timed out")
    except Exception as e:
        logger.error(f"Search failed: {e}")
        send_error_notification(str(e))


def send_error_notification(error_msg):
    """Send error notification to Slack"""
    try:
        from gowild_searcher import send_slack_notification
        
        error_summary = f"❌ *GoWild Search Failed*\n\nError: {error_msg}\nTime: {datetime.now().isoformat()}"
        send_slack_notification(error_summary)
    except Exception as e:
        logger.error(f"Failed to send error notification: {e}")


def start_scheduler():
    """Start the daily scheduler"""
    # Schedule for 6:00 AM daily
    schedule.every().day.at("06:00").do(run_search)
    
    logger.info("Scheduler started - runs daily at 6:00 AM")
    logger.info("Press Ctrl+C to stop")
    
    # Run initial search
    logger.info("Running initial search...")
    run_search()
    
    # Keep running
    while True:
        schedule.run_pending()
        time.sleep(60)  # Check every minute


if __name__ == '__main__':
    import argparse
    
    parser = argparse.ArgumentParser(description='GoWild Fare Scheduler')
    parser.add_argument('--start', '-s', action='store_true', help='Start daily scheduler')
    parser.add_argument('--run-now', '-r', action='store_true', help='Run search immediately')
    
    args = parser.parse_args()
    
    if args.start:
        start_scheduler()
    elif args.run_now:
        run_search()
    else:
        # Default: run once
        run_search()
